contributors/design-proposals/predicates-ordering.md

```diff
@@ -0,0 +1,90 @@
+# predicates ordering
+
+
+
+status: proposal
+
+Author: Yassine TIJANI.
+Approvers: 
+* gmarek
+* bsalamat
+* k82cn
+
+
+
+
+## Abstract
+
+This document describes how and why reordering predicates helps to achieve performance for the kubernetes scheduler. We will expose the motivations behind this proposal, The two steps/solution we see to tackle this problem and the timeline decided to implement these.
+
+
+## Motivation
+
+While working on a [Pull request](https://github.com/kubernetes/kubernetes/pull/50185) related to a proposal, we saw that the order of running predicates isn’t defined. 
+
+This makes the scheduler perform extra-computation that isn’t needed, As an example we [outlined](https://github.com/kubernetes/kubernetes/pull/50185) that the kubernetes scheduler runs predicates against nodes even if marked “unschedulable”.
+
+Reordering predicates allows us to avoid this problem, by computing the most restrictive predicates first. To do so, we propose two reordering types.
+
+
+
+## Static ordering
+
+This ordering will be the default ordering. If a policy config is provided with a subset of predicates, only those predicates will be invoked using the static ordering. 
+
+
+
+
+|Position                  | Predicate                        | comments (notes, justifications...)              |
+ ----------------- | ---------------------------- | ------------------
+| 1 | `CheckNodeConditionPredicate`  | we really don’t wanna check predicates against unschedulable nodes. |
```

Comment:
@yastij , as we talked on Slack, prefer to also add something like `O(n), n= # of node`. The user can evaluate the complexity of predicates based on their environment; so the lowest complexity with  most restrictive come first.

Reply:
@k82cn - I thought that time complexity might be more relevant in the user docs, as most of end-user will refer to the documentation when trying to check time complexity, you think we should add it here instead ?

Reply:
After thinking a bit more about time complexity, it seems that giving an estimation using the number of nodes might be not very useful, since the vast majority of predicates are computed against every node using the parallelized workqueue with a checkNode function.

@bsalamat @k82cn  - to your knowledge is there any common input to predicates that can gives us insights on time complexity ?

Reply:
I don't think there is any way to find runtime complexity of predicates other than examining their algorithms, but the autoscaler team recently ran some tests to find execution times of various predicates. You can find the results here: https://github.com/kubernetes/autoscaler/issues/257. It gives you some idea about the runtime complexity of the predicates.

Reply:
@bsalamat - thank you, indeed it should be enough